### ISSUE TITLE
[Flaky Test] [sig-apps] DisruptionController should block an eviction until the PDB is updated to allow it

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes/kubernetes#92977]
**Which jobs are flaking**: all of the e2e jobs that run this test

**Which test(s) are flaking**: `

`[sig-apps] DisruptionController should block an eviction until the PDB is updated to allow it`

**Testgrid link**: 

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=DisruptionController%20should%20block%20an%20eviction%20until%20the%20PDB%20is%20updated%20to%20allow%20it

https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5

**Reason for failure**:

mostly:
```
test/e2e/apps/disruption.go:273
Jul  9 23:34:27.116: Expected an error, got nil
test/e2e/apps/disruption.go:327
```

**Anything else we need to know**:
- links to go.k8s.io/triage appreciated
- links to specific failures in spyglass appreciated

/sig apps

xref: https://github.com/kubernetes/kubernetes/issues/92937

**Comment by BenTheElder** [comment]
approx: https://github.com/kubernetes/kubernetes/compare/70e09f2c2...82baa2690

**Comment by BenTheElder** [comment]
my suspicion: https://github.com/kubernetes/kubernetes/commit/9a23ee02b38bf367013532887c998adb80c0880a

**Comment by BenTheElder** [comment]
I can't find anything else that seems relevant in the time frame, and it is the only change to the test file in that time frame.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/91342 merged recently, not sure if it's relevant

**Comment by hakman** [comment]
Not sure what the procedure it, but we can revert #92745 or try some other image instead of Pause, if you think it helps.
I don't have a good explanation why the change would make the test flake like this.

**Comment by BenTheElder** [comment]
I don't have a good explanation, just that it affects the test and is in
the time frame that seems to fit the flake spike

It may take a bit to pin down what exactly happened here :/

I can't look for a little bit.

**Comment by hakman** [comment]
I prepared a PR to revert the change, just incase.

**Comment by hakman** [comment]
Good news and bad news kind of thing... 😞 
`[sig-apps] DisruptionController should block an eviction until the PDB is updated to allow it` fails in the revert PR also:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92982/pull-kubernetes-e2e-kind/1281849352104448001

**Comment by hakman** [comment]
I did an experiment reverting the second suspect in https://github.com/kubernetes/kubernetes/pull/92984. I couldn't reproduce the issue anymore even though I tried. You can check the results in the PR history.
https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=92984

**Comment by hasheddan** [comment]
@hakman I believe I have found the issue here. More information and fix incoming

**Comment by hakman** [comment]
Thanks for the update @hasheddan

**Comment by hasheddan** [comment]
I just about have a fix ready here. I was able to narrow the commit range to https://github.com/kubernetes/kubernetes/compare/82baa2690...dd649bb7e by looking at when this flake first occurred in https://testgrid.k8s.io/sig-release-master-blocking#kind-ipv6-master-parallel. It looks like @liggitt was correct that this was introduced in #91342. I believe what is happening is that in the test we are getting getting a `Pod` with a `DeletionTimestamp` set (the one that we previously successfully issued an eviction for) before waiting for 3 pods to be ready again in the namespace (https://github.com/kubernetes/kubernetes/blob/c2d15418316e9a02bf5692de80064556fb4f89f0/test/e2e/apps/disruption.go#L317). Before the aforementioned PR caused us to skip checking PDBs when `DeletionTimestamp` is set we would still report back that eviction was not possible because our `Pod` with `DeletionTimestamp` did not meet the other criteria (i.e. `Phase` was not `Succeeded`, `Failed`, or `Pending`), which caused us to still check PDBs, which would show us that we did not have the budget to delete (even though it was not really taking into account the `Pod` that we were actually attempting to evict).

I think there are a few things we can do to address this:
1. In `waitForPodsOrDie` we should check that not only is the `Pod` `Ready`, but also that it does not have `DeletionTimestamp` set.
2. Move `waitForPodsOrDie` above where we call `locateRunningPod` so that we do not attempt to get a `Pod` before all of the pods are ready.
3. Check in `locateRunningPod` that the `Pod` does not have a `DeletionTimestamp`.

Just doing (3) by itself would address our current problem, but (1) and (2) also seem to make sense in the context of what we are trying to check here.
